docs/architecture/adr-031-msg-service.md

```diff
@@ -0,0 +1,216 @@
+```protobuf
+message MsgSubmitProposal
+	option (cosmos_proto.msg_return) = “uint64”;
+	bytes delegator_address = 1;
+	bytes validator_address = 2;
+	repeated sdk.Coin amount = 3;
+}
+```
+
+This was never adopted, however, and we currently don’t have a mechanism for specifying
+the return type of a `Msg`.
+
+Having a well-specified return value for `Msg`s would improve client UX. For instance,
+in `x/gov`,  `MsgSubmitProposal` returns the proposal ID as a big-endian `uint64`.
+This isn’t really documented anywhere and clients would need to know the internals
+of the SDK to parse that value and return it to users.
+
+Also, there may be cases where we want to use these return values programatically.
+For instance, https://github.com/cosmos/cosmos-sdk/issues/7093 proposes a method for
+doing inter-module Ocaps using the `Msg` router. A well-defined return type would
+improve the developer UX for this approach.
+
+## Decision
+
+We decide to add support for using protobuf `service` definitions for `Msg`s.
+
+A `Msg` `service` definition would be written as follows:
+
+```proto
+package cosmos.gov;
+
+service Msg {
+  rpc SubmitProposal(MsgSubmitProposal) returns (MsgSubmitProposalResponse);
+}
+
+// Note that for backwards compatibility this uses MsgSubmitProposal as the request
+// type instead of the more canonical MsgSubmitProposalRequest
+message MsgSubmitProposal {
+  google.protobuf.Any content = 1;
+  bytes proposer = 2;
+}
+
+message MsgSubmitProposalResponse {
+  uint64 proposal_id;
+}
+```
+
+While this is most commonly used for gRPC, overloading protobuf `service` definitions like this does not violate
+the intent of the [protobuf spec](https://developers.google.com/protocol-buffers/docs/proto3#services) which says:
+> If you don’t want to use gRPC, it’s also possible to use protocol buffers with your own RPC implementation.
+With this approach, we would get an auto-generated `MsgServer` interface:
+
+In addition to clearly specifying return types, this has the benefit of generating client and server code. On the server
+side, this is almost like an automatically generated keeper method and could maybe be used intead of keepers eventually
+(see [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093)):
+
+```go
+type MsgServer interface {
+  SubmitProposal(context.Context, *MsgSubmitProposal) (*MsgSubmitProposalResponse, error)
+}
+```
+
+On the client side, developers could take advantage of this by creating RPC implementations that encapsulate transaction
+logic. Protobuf libraries that use asynchronous callbacks, like [protobuf.js](https://github.com/protobufjs/protobuf.js#using-services)
+could use this to register callbacks for specific messages even for transactions that span include multiple `Msg`s.
+
+For backwards compatibility, existing `Msg` types should be used as the request parameter
+for `service` definitions. Newer `Msg` types which only support `service` definitions
+should use the more canonical `Msg...Request` names.
```

Comment:
For consistency, may be we can avoid `Request` suffix for newer `Msg`'s as well

Reply:
I guess that's okay although I prefer to keep the buf linter defaults where possible.

Reply:
I would make this case a buf linter exception, for the sake of consistency, and also `MsgAbc` reads better than `MsgAbcRequest` imo.

Reply:
Okay, I guess we can enable a linter exception.